pulp_container/tests/functional/api/test_push_content.py

```diff
@@ -6,49 +6,186 @@
-        distribution = distributions_api.list(name="foo/bar").results[0]
-        self.addCleanup(distributions_api.delete, distribution.pulp_href)
+            self.registry.push(local_url)
+
+    def test_push_with_dist_perms(self):
+        """
+        Test that it's enough to have container distribution and namespace perms to perform push.
+
+        It also checks read abilities for users with different set of permissions.
+        """
+        repo_name = "test/perms"
+        local_url = "/".join([self.registry_name, f"{repo_name}:2.0"])
+        image_path = f"{DOCKERHUB_PULP_FIXTURE_1}:manifest_a"
+        self._push(image_path, local_url, self.user_creator)
+
+        self.assertEqual(self.pushrepository_api.list(name=repo_name).count, 1)
+        self.assertEqual(self.user_creator["pushrepository_api"].list(name=repo_name).count, 1)
+        self.assertEqual(self.user_collaborator["pushrepository_api"].list(name=repo_name).count, 1)
+        self.assertEqual(self.user_reader["pushrepository_api"].list(name=repo_name).count, 1)
+        self.assertEqual(self.user_helpless["pushrepository_api"].list(name=repo_name).count, 0)
+
+        # cleanup, namespace removal also removes related distributions
+        namespace = self.namespace_api.list(name="test").results[0]
+        self.addCleanup(self.namespace_api.delete, namespace.pulp_href)
+
+    def test_push_with_view_perms(self):
+        """
+        Test that push is not working if a user has only view permission for push repos.
+        """
+        repo_name = "unsuccessful/perms"
+        local_url = "/".join([self.registry_name, f"{repo_name}:2.0"])
+        image_path = f"{DOCKERHUB_PULP_FIXTURE_1}:manifest_a"
+        with self.assertRaises(exceptions.CalledProcessError):
+            self._push(image_path, local_url, self.user_reader)
+
+    def test_push_with_no_perms(self):
+        """
+        Test that user with no permissions can't perform push.
+        """
+        repo_name = "unsuccessful/perms"
+        local_url = "/".join([self.registry_name, f"{repo_name}:2.0"])
+        image_path = f"{DOCKERHUB_PULP_FIXTURE_1}:manifest_a"
+        with self.assertRaises(exceptions.CalledProcessError):
+            self._push(image_path, local_url, self.user_helpless)
+
+    def test_push_to_existing_namespace(self):
+        """
+        Test the push to existing namespace with collaborator permissions.
+
+        Container distribution perms and manage-namespace one should be enough.
+        """
+        repo_name = "team/owner"
+        local_url = "/".join([self.registry_name, f"{repo_name}:2.0"])
+        image_path = f"{DOCKERHUB_PULP_FIXTURE_1}:manifest_a"
+        self._push(image_path, local_url, self.user_creator)
+
+        collab_repo_name = "team/owner"
+        local_url = "/".join([self.registry_name, f"{collab_repo_name}:2.0"])
+        image_path = f"{DOCKERHUB_PULP_FIXTURE_1}:manifest_b"
+        self._push(image_path, local_url, self.user_collaborator)
```

Comment:
:+1:

Reply:
I am wondering if the manage_namespace permission alone should do the trick here.
Sure it would be better to have it as a object permission. But that shouldn't block us to get this PR merged.